### ISSUE TITLE
supervisorctl reload failed restarting supervisord

### ISSUE THREAD

**Issue by laurent-claustre** [tiagocoutinho/multivisor#12]
Hi Tiago,

testing with the examples you provide, supervisorctl reload did crash the supervisord deamon.
I did start supervisord in foreground and get the following:

No handlers could be found for logger "MVRPC"
2019-02-05 10:35:11,607 INFO stopped: vacuum_OH2_1 (terminated by SIGTERM)
2019-02-05 10:35:11,607 INFO stopped: vacuum_OH2_2 (terminated by SIGTERM)
2019-02-05 10:35:11,607 INFO stopped: vacuum_OH1 (terminated by SIGTERM)
2019-02-05 10:35:11,609 INFO stopped: wcid00b (terminated by SIGTERM)
2019-02-05 10:35:11,609 INFO stopped: wcid00a (terminated by SIGTERM)
2019-02-05 10:35:11,609 INFO stopped: wcid00c (terminated by SIGTERM)
2019-02-05 10:35:11,609 INFO stopped: wcid00d (terminated by SIGTERM)
2019-02-05 10:35:11,611 INFO stopped: exits_10s (terminated by SIGTERM)
2019-02-05 10:35:11,629 INFO RPC interface 'supervisor' initialized
2019-02-05 10:35:11,629 INFO RPC interface 'multivisor' initialized
2019-02-05 10:35:11,629 CRIT Server 'inet_http_server' running without any HTTP authentication checking
2019-02-05 10:35:11,629 INFO supervisord started with pid 7157
Assertion failed: nbytes == sizeof dummy (src/signaler.cpp:236)
Aborted

my python installation is as following:
miniconda 4.6.2
python 2.7.15 
multivisor 5.0.0 installed with pip

**Comment by laurent-claustre** [comment]
sorry, i forget to tell you, removing the entry [rpcinterface:multivisor] fix the issue.

**Comment by tiagocoutinho** [comment]
Thanks for the report @laurent-claustre. Looks like a problem in multivisor rpc interface.
I am having a look at it

**Comment by tiagocoutinho** [comment]
Bang! I can reproduce it. Keep you informed of my progress.
Looks like a problem in either zeromq or zerorpc.

**Comment by tiagocoutinho** [comment]
Regarding the supervisord crash:
The multivisor RPC runs in a separate thread because it is gevent based. When you hit reload, the multivisor RPC along with its thread are destroyed and a new one is created. 
As far as I could see I seems to be the destruction that causes the problem.

So far I have failed to reproduce the problem with a simple zerorpc server.

We have two options:
1. internally, when supervisord is reloaded, we prevent the destruction of the multivisor RPC object and the thread. Instead we just update the reference to the new supervisor object. This defeats a little bit the purpose of making a full reload but it should fix the problem
2. Dig a little deeper and see what we find.

I propose I implement option 1 in a version 5.0.1 probably together with a fix for the stop. In parallel I continue to investigate what is really going on.
This seems the quickest way to unblock you.
Do you agree?

Thanks for reporting and the patience ;-)

**Comment by laurent-claustre** [comment]
Hi Tiago, ok for option 1, I need a first fix in order to get our ansible recipe accepted  by our team.
Thanks a lot for your help.

**Comment by tiagocoutinho** [comment]
Here is what I found so far:
1. when you run `reload` in *supervisorctl* actually it invokes a `restart` in *supervisord* (don't ask me why :-)
2. seems zerorpc server does not restart properly. I created an issue (see 0rpc/zerorpc-python#208)
3. when you ask *supervisord* to restart, it closes **all** file descriptors from 5..1024. This is absolutely insane. This means it also kills the file descriptors needed by gevent machinery and 0MQ socket that is used by zerorpc

All these things put together account for the bug we see.

The first patch I am working simply prevents *supervisord* from closing the file descriptors.
It is not ideal but it seems to work. I guess I will make a 5.0.1 with this fix until we hear more news from ZeroRPC.

Concerning the STOP problem I was able to see it once while debugging the *reload* problem but so far I am unable to reproduce it. I will keep you informed when I release 5.0.1 and also of my progress with the STOP issue

**Comment by laurent-claustre** [comment]
hummm, I understand in the meantime we could disable reloading supervisord in our ansible recipe till you can provide the release 5.0.1.
Thanks a lot again.

**Comment by tiagocoutinho** [comment]
@laurent-claustre ,I just released 5.0.1. 
Could you have a try and let me know if
* the restart problem is fixed
* the stop problem is fixed

Thanks in advance

**Comment by laurent-claustre** [comment]
I did test it, reload is no more crashing supervisord.
But, now multivisor complains about 'lost connection'
60.103.228.145 - - [2019-02-12 19:17:41] "GET /data HTTP/1.1" 200 324 0.000946
INFO 2019-02-12 19:17:49,497 multivisor.lmpx4: Lost remote
INFO 2019-02-12 19:17:49,497 multivisor.lmpx4: (re)initializing...
INFO 2019-02-12 19:18:04,501 multivisor.lmpx4: Lost remote
INFO 2019-02-12 19:18:04,501 multivisor.lmpx4: (re)initializing...

my multivisor config file:
[supervisor:lmpx4]
url=lmpx4:9002

[global]
name=Test

and my supervisor config file to run multivisor:
program:multivisor]
command=bash -c "source /users/blissadm/conda/miniconda/bin/activate multivisor && exec multivisor --bind=0:22000 -c /users/blissadm/local/daemon/config/multivisor.conf"
user=blissadm
startsecs=2
autostart=true
redirect_stderr=true
stdout_logfile=/var/log/%(program_name)s.log
stdout_logfile_maxbytes=1MB
stdout_logfile_backups=10
stdout_capture_maxbytes=1MB

In case of missconfiguration with supervisor, this is its configuration under /etc/supervisor/supervisor.conf:
; supervisor config file
[inet_http_server]
port = *:9001

[supervisord]
logfile=/var/log/supervisor/supervisord.log ; (main log file;default $CWD/supervisord.log)
pidfile=/var/run/supervisord.pid ; (supervisord pidfile;default supervisord.pid)
childlogdir=/var/log/supervisor            ; ('AUTO' child log dir, default $TEMP)
identifier = lmpx4 
; new rpc interface for multivisor 
[rpcinterface:multivisor]
supervisor.rpcinterface_factory = multivisor.rpc:make_rpc_interface
bind=*:9002

[supervisorctl]
serverurl=http://localhost:9001

[include]
files = /etc/supervisor/conf.d/*.conf /etc/supervisor/bliss.d/*.conf

**Comment by tiagocoutinho** [comment]
In which host are you running multivisor? Is it lmpx4?
Did you update multivisor on all hosts?

**Comment by laurent-claustre** [comment]
yes on lmpx4, only, ok tomorrow I will make a try with just your example of conf.

**Comment by tiagocoutinho** [comment]
I had a look at lmpx4 logs. It looks strange.
It is as if the RPC is not accepting any connections
Do you mind if I login to lmpx4 tomorrow and make a few tests?

**Comment by laurent-claustre** [comment]
No problem Tiago, you can make some tests.

**Comment by laurent-claustre** [comment]
did you fix something in multivisor env in blissadm account? now it works fine

**Comment by laurent-claustre** [comment]
ok, i think the problem was when i did add the multivisor.conf file into ~blissadm/local/daemon/config/supervisor.d/ and use supervisorctl reload (again). multivisor is well started but complained about 'lost connection'. To fix the problem I did restart supervisor with the service command (service supervisor restart). 
Anyway, we can leave with service command (need to be root of course) in the ansible recipe.
For users the supervisorctl reload must work when you want to update a configuration file like to add a new server.

**Comment by tiagocoutinho** [comment]
I logged in to lmpx4. I noticed the problem you mentioned. I added some logging and then I tried to restart supervisor with `service supervisor restart`. After that the problem disappeared.

> For users the supervisorctl reload must work when you want to update a configuration file like to add a new server.
After reload with supervisorctl are things still working?

Hint: you can use multivisor web page, in the supervisor tab, each panel as a menu where you can "restart" the respective supervisor. It does the same as the reload command of supervisorctl

**Comment by laurent-claustre** [comment]
>  reload with supervisorctl are things still working?
yes.
Think for the hint.
A last little thing, with supervisorctl you can tail on the stdout of the process, within multivisor web if I click for instance my maxipix icon to get log, the log page is empty? Did I missconfigure something?


>

**Comment by tiagocoutinho** [comment]
> A last little thing, with supervisorctl you can tail on the stdout of the process, within multivisor web if I click for instance my maxipix icon to get log, the log page is empty? Did I missconfigure something?
No, it is another bug I noticed already. I am working on it right now :-)

**Comment by tiagocoutinho** [comment]
I created a new issue for this #14.
I already found the problem. Building a 5.0.2 soon!

**Comment by tiagocoutinho** [comment]
If you agree I can close this issue

**Comment by laurent-claustre** [comment]
ok I can even close it myself. and I will download the 5.0.2 asap it is available.
Many many thanks for the fixes
